lib/ansible/modules/storage/zfs/zfs_facts.py

```diff
@@ -0,0 +1,198 @@
+        description:
+            - Specifies if property values should be displayed in machine
+              friendly format.
+        type: bool
+        default: False
+        required: false
+    properties:
+        description:
+            - Specifies which dataset properties should be queried in comma-separated format.
+              For more information about dataset properties, check zfs(1M) man page.
+        alias: [ "props" ]
+        type: str
+        default: all
+        required: false
+    type:
+        description:
+            - Specifies which datasets types to display. Multiple values have to be
+              provided in comma-separated form.
+        alias: [ "props" ]
+        type: str
+        default: all
+        choices: [ 'all', 'filesystem', 'volume', 'snapshot', 'bookmark' ]
+        required: false
+    depth:
+        description:
+            - Specifiies recurion depth.
+        type: int
+        default: None
+        required: false
+'''
+
+EXAMPLES = '''
+# Gather facts about ZFS dataset rpool/export/home
+zfs_facts: dataset=rpool/export/home
+
+# Report space usage on ZFS filesystems under data/home
+zfs_facts: name=data/home recurse=yes type=filesystem
+debug: msg='ZFS dataset {{ item.name }} consumes {{ item.used }} of disk space.'
+with_items: '{{ ansible_zfs_datasets }}
+'''
+
+RETURN = '''
+'''
+
+import os
+from collections import defaultdict
+
+
+SUPPORTED_TYPES = ['all', 'filesystem', 'volume', 'snapshot', 'bookmark']
+
+
+class ZFSFacts(object):
+    def __init__(self, module):
+
+        self.module = module
+
+        self.name = module.params['name']
+        self.recurse = module.params['recurse']
+        self.parsable = module.params['parsable']
+        self.properties = module.params['properties']
+        self.type = module.params['type']
+        self.depth = module.params['depth']
+
+        self._datasets = defaultdict(dict)
+        self.facts = []
+
+    def dataset_exists(self):
+        cmd = [self.module.get_bin_path('zfs')]
+
+        cmd.append('list')
+        cmd.append(self.name)
+
+        (rc, out, err) = self.module.run_command(cmd)
+
+        if rc == 0:
+            return True
+        else:
+            return False
+
+    def get_facts(self):
+        cmd = [self.module.get_bin_path('zfs')]
+
+        cmd.append('get')
+        cmd.append('-H')
+        if self.parsable:
+            cmd.append('-p')
+        if self.recurse:
+            cmd.append('-r')
+        if int(self.depth) != 0:
+            cmd.append('-d')
+            cmd.append('%s' % self.depth)
+        if self.type:
+            cmd.append('-t')
+            cmd.append(self.type)
+        cmd.append('-o')
+        cmd.append('name,property,value')
+        cmd.append(self.properties)
+        cmd.append(self.name)
+
+        (rc, out, err) = self.module.run_command(cmd)
+
+        if rc == 0:
+            for line in out.splitlines():
+                dataset, property, value = line.split('\t')
```

Comment:
Does line parsing depend on presence on self.parsable? BTW, why could we need non-parsable output?